File: python/strings/group_anagram.py

def groupAnagrams(strs):
    def rearrange_string(s):
        s = list(s)
        s.sort()

        return ''.join(s)

    # sortedString: [array of inputted string]
    lookup = {}

    for s in strs:
        _s = rearrange_string(s)

        if _s in lookup:
            lookup[_s] += [s]
        else:
            lookup[_s] = [s]

    return list(lookup.values())


strs = ["eat", "tea", "tan", "ate", "nat", "bat"]

print(groupAnagrams(strs))

File: random/groupAnagram.ts
function groupAnagrams(strs: string[]): string[][] {
    let strsMapping: Record<string, Array<string>> = {};

    for (const str of strs) {
        const sortedStr = [...str].sort((a, b) => a.localeCompare(b)).join('');
        const prev = strsMapping[sortedStr];

        if (prev) {
            strsMapping[sortedStr] = [...prev, str];
        } else {
            strsMapping[sortedStr] = [str];
        }
    }

    return Object.values(strsMapping);
};
